### DISCUSSION TITLE
How to use non-mapped values for db statements

### DISCUSSION THREAD

**Discussion by jenstroeger** [sqlalchemy/sqlalchemy#10882] [Usage Questions] [answered]
Suppose I have two decimal values that I’d like to add using Postgres. The raw SQL would be something like:
```
postgres=# select 42.42::numeric + 24.24::numeric;
 ?column? 
----------
    66.66
(1 row)
```
I’d like to achieve that using SQLA, but can’t quite settle on the correct (recommended) approach. So, I can either
```python
a = decimal.Decimal("42.42")
b = decimal.Decimal("24.24")

result = session.scalars(
    sa.select(sa.sql.expression.literal_column(str(a), Numeric) + sa.sql.expression.literal_column(str(b), Numeric)
).one()
```
or
```python 
result = session.scalars(
    sa.select(sa.cast(str(a), Numeric) + cast(str(b), Numeric)
).one()
```
Is this the correct way of delegating the calculation to the db, or is there another recommended/more efficient one? Is there a way around using `str()` here at all?

**Comment by CaselIT** [comment]
Hi,

I don't think you should use str().
You seem to want bindparam instead of cast or literal_column

**Reply by jenstroeger** [reply]
Thank you, @CaselIT, looks like [bindparam](https://docs.sqlalchemy.org/en/20/core/sqlelement.html#sqlalchemy.sql.expression.bindparam) worked:
```python
result = session.scalars(
    select(cast(bindparam("a"), Numeric) + cast(bindparam("b"), Numeric),
    {"a": a, "b": b}
).one()
```
Do I need the cast here? I suppose I do to ensure that PG interprets the arguments as `numeric` types?

**Reply by CaselIT** [reply]
I actually forgot about `literal`. So your query could be rendered as `select(literal(a)+literal(b))`

It seems that the cast is not used, at least in psycopg2, but the returned value is a decimal so it should have used the correct types in the db. If you want to make sure of the type adding a cast is safer.

**Reply by jenstroeger** [reply]
Hmm ok, so that would make it
```python
result = session.scalars(
    select(literal(a, Numeric) + literal(b, Numeric)),
).one()
```
If I understand the [docs](https://docs.sqlalchemy.org/en/20/core/sqlelement.html#sqlalchemy.sql.expression.literal) then the second `type_` arg should ensure the `numeric` type on the db, though the compiled query is…
```
SELECT 42.42 + 24.24 AS anon_1
```
Adding [`cast()`](https://docs.sqlalchemy.org/en/20/core/sqlelement.html#sqlalchemy.sql.expression.cast) around `literal()` compiles to
```
SELECT CAST(42.42 AS NUMERIC) + CAST(24.24 AS NUMERIC) AS anon_1 
```
which is what I want.

So what’s the suggestion then, what’s the difference/better recommended — `bindparam()` or `literal()`?

**Reply by CaselIT** [reply, marked as answer]
literal is actually just a constructor that returns bindparam with an anonymous name so they are the same exact thing.

Setting the type on literal / bindparam does not render a cast (usually, it may do so in some cases in pg) it's just inform sqlalchemy and the driver of what type it should use to avoid the guessing that does sqlalchemy. So if you want a cast do use it explicitly. (a maybe more convenient way to add the cast by calling `.cast(...)` on the literal: `select(literal(a).cast(Numeric) + literal(b).cast(Numeric))`

**Reply by jenstroeger** [reply]
Great, thanks @CaselIT, worked like a charm!

(Appreciate the hint on [internal workings](https://github.com/sqlalchemy/sqlalchemy/blob/8f4ac0c0f07509d2f8a4bce9cbb07ac08ad04044/lib/sqlalchemy/sql/elements.py#L155), interesting to follow the call chain 🤓)
